mdsuite/calculators/structure_factor.py

```diff
@@ -99,246 +110,264 @@ class StructureFactor(Calculator):
+
+        # set args that will affect the computation result
+        self.args = Args(
+            number_of_bins=self.rdf_data.computation_parameter["number_of_bins"],
+            cutoff=self.rdf_data.computation_parameter["cutoff"],
+            number_of_configurations=self.rdf_data.computation_parameter[
+                "number_of_configurations"
+            ],
+            resolution=resolution,
+        )
+        self.q_values = np.linspace(0.5, 12, resolution)
+
+        self._compute_angstsrom_volume()  # correct the volume of the system.
+        self.number_of_atoms = sum(
+            [species.n_particles for species in self.experiment.species.values()]
+        )
+        self.total_density = self.number_of_atoms / self.volume
+
+        # Construct the species dictionary and fill all species information.
+        self.species_dict = {}
+        for species in self.experiment.species:
+            n_particles = self.experiment.species[species].n_particles
+            particle_density = n_particles / self.volume
+            molar_fraction = n_particles / self.number_of_atoms
+            self.species_dict[species] = SpeciesData(
+                particle_density=particle_density,
+                molar_fraction=molar_fraction,
             )
+        self._compute_form_factors()
 
-            if self.load_data:
-                out[self.experiment.name] = self.experiment.export_property_data(
-                    {"Analysis": self.analysis_name}
-                )
-            else:
-                out[self.experiment.name] = self.run_analysis()
+    def _compute_angstsrom_volume(self):
+        """
+        Compute the volume of the box in Ansgtrom.
 
-        if len(self.experiments) > 1:
-            return out
-        else:
-            return out[self.experiment.name]
+        The data for scattering features is in Angstrom. Furthermore, the radial
+        component of the calculator is integrated over.
 
-    @staticmethod
-    def gauss(a, b, scattering_scalar):
+        Returns
+        -------
+        Updates the volume attribute of the class.
         """
-        Calculates the gauss functions that are required for the atomic form factors
+        volume_si = self.experiment.volume * (self.experiment.units["length"] ** 3)
+
+        self.volume = volume_si / 1e-10 ** 3
+
+    def _compute_form_factors(self):
         """
-        return a * np.exp(-b * (scattering_scalar / (4 * np.pi)) ** 2)
+        Compute the atomic form factors for each species and add them to the
+        data class.
 
-    def atomic_form_factors(self, scattering_scalar):
+        Returns
+        -------
+        Updates the data class for each species.
         """
-        Calculates the atomic form factors for all elements in the species
-        dictionary and returns it
+        for name, species_data in self.species_dict.items():
+            aff_data = self.form_factor_data.loc[self.form_factor_data["Element"] == name]
```

Comment:
maybe change `aff_data` to something more descriptive. But I don't know if it is already descriptive for most people it would be fine.

Reply:
For this one, in another method I wrote it in the doc string as there is not reasonable way of writing `atomic_form_factor_data` with such a long assignment. My though here for additional clarity would be to add it to the doc-string and add a comment over its first assignment along the lines of:

```python
# aff -> atomic form factor
aff_data = ...
```

Reply:
sounds good